### PR TITLE
Project and unit field integrations

### LINKED ISSUE #142: Field additions on project & unit levels to export entire development projects (projects/units/properties)
**Problem description**
So far, there's a minimal number of fields available in SwissRETS for the export of entire development projects (consisting of multiple units / multiple properties per unit). 

See: 
https://swissrets.ch/docs/noNamespace/element/export.projects.html
https://github.com/qualipool/swissrets/blob/master/examples/full.xml

**Desired Solution**
Additional fields to export basic informations about the development project / units in order for publishers (real estate classifieds/marketplaces and/or individual webpages) to display the information properly. 

For an MVP version we suggest the following additions: 

Project level: 
- Project name
- Project short description
- Project location
- Project logo
- Construction status
- Total number of residential units
- Available from
- Price from
- Price to
- M2 Livingspace from
- M2 Livingspace to
- Number of rooms from
- Number of rooms to
- Project photos
- Dedicated project website link
- Project videos link (youtube, or others)
- Developer description
- Developer logo
- Developer image

Unit level: 
- Unit name
- Total number of floors

### LINKED ISSUE #162: The current structure allows for properties to be connected to multiple projects/units
Small concern that I just realised.

The current structure allows for properties to be connected to multiple projects/units (manyToMany) This might actually make things a bit complicated. We should think about putting the reference to a unit within the property itself instead and remove the "properties" section here. That way a property can only belong to one unit and this would make things a little more restrictive and simpler. Unless there is a use case for a property to exist within multiple projects/units?

But I guess this would require an alteration of the existing schema and would be a different issue. I can create one if that is a concern that is shared, and not just my paranoia speaking :-) .

Example
```xml
<properties>
  <property>
    <unitId>{ID of Unit}</unitId>
```

_Originally posted by @jensstalder in https://github.com/qualipool/swissrets/issues/142#issuecomment-756032061_

### LINKED ISSUE #174: TBD: Explicit Deletion of Properties
**Problem description**

Currently a realtor has to provides all his properties within one single XML-file. The distinction whether a property should be deleted happens implicitly: If the realtor, in comparison to his previous import, leaves out one or more properties, those properties must be deleted automatically.

This implies that each provider has to check the delta and remove previously imported listings.

**Desired Solution**

This issue won't provide any solution to this topic, the purpose is to open a discussion between stakeholders, if this workflow is the desired one or if the SwissRets standard could benefit from some sort of CRUD-attributes.

Just to clarify this issues by a simple example:

```
<?xml version="1.0" standalone="yes"?>
<xs:schema>
  <xs:simpleType name="propertyStatus">
    <xs:annotation>
      <xs:documentation>Status that indicates the desired status of a property.</xs:documentation>
    </xs:annotation>
    <xs:restriction base="xs:string">
      <xs:enumeration value="CREATE-UPDATE" />
      <xs:enumeration value="DELETE" />
    </xs:restriction>
  </xs:simpleType>
  [...]
</xs:schema>
```

### PR DESCRIPTION
**Pull request by jensstalder**
## Semantic versioning

This PR is a **Minor** (It was aggreed that the change will not affect 3rd party folks since it has not been used yet)

## Generall things to consider 
- I have read the [**CONTRIBUTING.md**](https://github.com/qualipool/swissrets/blob/master/CONTRIBUTING.md) document.
- If my change requires documentation, I added it already.
- I have added tests to cover all my changes.
- All new and existing tests pass.

## List of issues related to this PR
- closes #162
- closes #142

### Added
- Fields to project and units

### Changed
- Direction of relationships between property and units

### CODE REVIEW

**Review by DrMueller — APPROVED** [review]

File: schema/schema.xsd [hunk 1]
```diff
@@ -1173,6 +1193,222 @@
...
+                              <xs:maxLength value="3" />
+                              <xs:pattern value="[A-Z]{3}" />
+                            </xs:restriction>
+                          </xs:simpleType>
+                        </xs:attribute>
+                      </xs:complexType>
+                    </xs:element>
+                    <xs:element name="characteristics" minOccurs="0" maxOccurs="1">
+                      <xs:annotation>
+                        <xs:documentation>Price ranges for marketing purposes.</xs:documentation>
```
Comment (jensstalder) [inline schema/schema.xsd:1287]: Wrong documentation. Looks like a copy paste error.

**Comment by boeschenstein** [comment]
Although it is not written in the docs yet (and this might be our own understanding), in general, ID fields are technical fields (from the source system), and "Reference Id" are logical fields. Therefore, I suggest renaming the new field from "projectUnitId" to "projectUnitReferenceId", so is clear that the link goes by project-unit-referenceId (not project-unit-id).

Please add information about the Id fields to the docs, so the usage is getting clearer.

**Comment by jensstalder** [comment]
My hope was to use the attribute id= so that we could make use of validation. similar to how html uses id attributes together with anchors for example. But I could not make this work. This would however require for the reference entities to be present in the XML document to make it valid. There are pros and cons to that. What do you think?

When I google this topic it's a bit weird.
* https://www.oreilly.com/library/view/xml-pocket-reference/9780596100506/ch01s02s12.html 
* https://www.data2type.de/xml-xslt-xslfo/xml-schema/datentypen-referenz/xs-id/
* https://www.w3.org/2002/ws/databinding/examples/6/09/Id/

Maybe you can find something more legible? 

This is what is currently documented:
```
id
          this ID needs to be unique and represents the id from the exporter (this can be the same as referenceId if the exporter is the original creator)  
          
referenceId
          Can be used to track the original creation id (this should usually be unique to the property) 

visualReferenceId
          Can should only be used for presentational purposes publishers tend to present the referenceId if this one is missing 
```
It does have implicit meaning that the "id" or more of a technical/structural id but that a consumer/importer should look at referenceId for import duplication checking instead. But internal referencing would actually use the id attribute if we want to go the "xml way" route.

**Comment by jensstalder** [comment]
Ok I got it working with the unit reference. The only thing that was missing. I also added examples and a test for this. See `examples/project-tree.xml`.

The only change is that I renamed the reference element to simply `<unitId>` and it now checks that it is valid.

I checked and it would also be possible to connect to `<unit><referenceId>` instead, but this is up for debate still? What do you guys think. Should we reference `id=""` or `<referenceId>`?

**Comment by jensstalder** [comment]
`<xs:element name="availability" minOccurs="1" maxOccurs="1">` For projects are "extending" the availability from properties. Which in my opinion does not make sense, since a project can't be "reserved" or "in-acquisition" or "taken". So we should change that enum to just have "active" for the moment. And we can add more in the future such as "private" if they make sense for projects.

**Pull request merged**